### DISCUSSION TITLE
useEventListener TS error

### DISCUSSION THREAD

**Discussion by aborodovisin** [juliencrn/usehooks-ts#653] [Help]
There is a way to directly enhance the `useEventListener` to support `CustomEvents`. I don't like so much the boilerplate code improving the global definition to bypass TypeScript. Seems this custom hooks missed support for `CustomEvents`?

**Comment by tamish-max** [comment]
I think this is mostly a **TypeScript typing limitation**, not a runtime limitation in the hook.

`useEventListener` is typed against the built-in event maps (`WindowEventMap`, `DocumentEventMap`, etc.), so for `CustomEvent`s the idiomatic TS solution is usually to augment the relevant event map:

```ts
declare global {
  interface WindowEventMap {
    'my-event': CustomEvent<{ foo: string }>
  }
}

useEventListener('my-event', event => {
  console.log(event.detail.foo)
})
```

So yes: the hook can *work* with custom events, but to get proper typing you typically need to extend the global map (or wrap the hook in your own typed helper).

If the project wants first-class CustomEvent ergonomics, it would probably need extra overloads/generics for `EventTarget`-style maps.
